### PR TITLE
Feature to support ObjLoader Zbrush MRGB 

### PR DESCRIPTION
**Pull request by Moriyer**
https://forum.babylonjs.com/t/feature-for-obj-loader-zbrush-mrgb-support/53190/3

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15636/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15636/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15636/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15636/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15636/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15636/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15636/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15636/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15636/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15636/merge/testResults/webgl2playwright/index.html

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 1]
```diff
@@ -590,14 +628,13 @@ export class SolidParser {
...
         const lines = lineLines.flat();
-
         // Look at each line
         for (let i = 0; i < lines.length; i++) {
             const line = lines[i].trim().replace(/\s\s/g, " ");
             let result;
-
             // Comment or newLine
             if (line.length === 0 || line.charAt(0) === "#") {
+                Logger.Log("Unhandled expression at line : " + line);
```
Comment (RaananW) [inline packages/dev/loaders/src/OBJ/solidParser.ts:637]: why is that needed?

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 2]
```diff
@@ -275,6 +276,17 @@ export class SolidParser {
...
 
+    /**
+     * To get color between color and extension color
+     * @param index Integer The index of the element in the array
+     * @returns value of target color
+     */
+    private _getColor(index: number) {
+        if (!this._loadingOptions.importVertexColors) return undefined;
+        if (this._extColors.length > 0) return this._extColors[index];
+        return this._colors[index];
```
Comment (RaananW) [inline packages/dev/loaders/src/OBJ/solidParser.ts:287]: no need to actually test for length > 0
```suggestion
        if (this._loadingOptions.importVertexColors) {
            return this._extColors[index] ?? this._colors[index];
        }
```

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 1 above)
Comment (Moriyer) [reply packages/dev/loaders/src/OBJ/solidParser.ts]: This my mistake, it should be delete

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 3]
```diff
@@ -275,6 +276,17 @@ export class SolidParser {
         //Pattern5 => triangle = ["-1/-1/-1","-2/-2/-2","-3/-3/-3","-1/-1/-1","-3/-3/-3","-4/-4/-4"];
     }
 
+    /**
+     * To get color between color and extension color
+     * @param index Integer The index of the element in the array
+     * @returns value of target color
+     */
+    private _getColor(index: number) {
+        if (this._loadingOptions.importVertexColors) {
```
Comment (RaananW) [inline packages/dev/loaders/src/OBJ/solidParser.ts:285]: Sorry, i sometimes forget we don't have noImplicitReturns really. You will need to add `else { return; }` to make typescript happy...

or something like this:

```javascript
return  this._loadingOptions.importVertexColors ? this._extColors[index] ?? this._colors[index] : undefined;
```

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 4]
```diff
@@ -537,6 +551,28 @@ export class SolidParser {
...
+        if (!regArray || regArray.length % 8 !== 0) {
+            return null;
+        }
+        const array = [];
+        for (let regIndex = 0; regIndex < regArray.length / 8; regIndex++) {
+            const m = regArray[regIndex * 8 + 0] + regArray[regIndex * 8 + 1];
+            const r = regArray[regIndex * 8 + 2] + regArray[regIndex * 8 + 3];
+            const g = regArray[regIndex * 8 + 4] + regArray[regIndex * 8 + 5];
+            const b = regArray[regIndex * 8 + 6] + regArray[regIndex * 8 + 7];
+            const color = { m: parseInt(m, 16) / 255, r: parseInt(r, 16) / 255, g: parseInt(g, 16) / 255, b: parseInt(b, 16) / 255 };
```
Comment (ryantrem) [inline packages/dev/loaders/src/OBJ/solidParser.ts:570]: Is the `m` property used anywhere?

File: packages/dev/loaders/src/OBJ/solidParser.ts [hunk 5]
```diff
@@ -537,6 +551,28 @@ export class SolidParser {
...
+        if (!line.startsWith("mrgb")) return null;
+        line = line.replace("mrgb", "").trim();
+        // if include vertex color , not load mrgb anymore
+        if (notParse) return [];
+        const regex = /[a-z0-9]/g;
+        const regArray = line.match(regex);
+        if (!regArray || regArray.length % 8 !== 0) {
+            return null;
+        }
+        const array = [];
```
Comment (ryantrem) [inline packages/dev/loaders/src/OBJ/solidParser.ts:564]: This is an `any` array. Can we make this either `{r: number, g: number, b: number}[]` or just directly construct Babylon `Color4` objects here (e.g. `Color4[]`)?

File: packages/dev/loaders/src/OBJ/solidParser.ts (hunk 4 above)
Comment (Moriyer) [reply packages/dev/loaders/src/OBJ/solidParser.ts:570]: m is not used, m is materialId here, I add it to keep all data save. Should I delete it?

**Review by ryantrem — APPROVED** [review]

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15636/merge/testResults/webgpuplaywright/index.html
